File: minimumPathSum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if len(grid) == 0: return 0
        m, n = len(grid), len(grid[0])
        for i in xrange(m):
            for j in xrange(n):
                if i == 0 and j==0 :
                    continue
                upcell   = grid[i-1][j] if i-1>=0 else sys.maxint
                leftcell = grid[i][j-1] if j-1>=0 else sys.maxint
                grid[i][j] += min(upcell, leftcell)
        return grid[m-1][n-1]

File: minimumPathSum.kt
class Solution {
    fun minPathSum(grid: Array<IntArray>): Int {
        val m = grid.size
        val n = grid[0].size
        val res = IntArray(n){0}
        for(i in 0 until m){
            for(j in 0 until n){
                if(i == 0 && j == 0){
                    res[j] = grid[i][j]
                } else if(i == 0){
                    res[j] = res[j-1] + grid[i][j]
                } else if(j == 0){
                    res[j] = res[j] + grid[i][j]
                } else{
                    if(res[j-1] < res[j]) res[j] = res[j-1] + grid[i][j]
                    else res[j] = res[j] + grid[i][j]
                }
            }
        }
        return res[n-1]
    }
}
